sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -559,27 +560,39 @@ public HttpPipeline getHttpPipeline() {
      */
-    public Mono<Void> downloadToFile(String filePath, BlobRange range, Integer blockSize,
-        ReliableDownloadOptions options, BlobAccessConditions accessConditions, boolean rangeGetContentMD5) {
-        return withContext(context -> downloadToFile(filePath, range, blockSize, options, accessConditions,
-            rangeGetContentMD5, context));
+    public Mono<Response<BlobProperties>> downloadToFileWithResponse(String filePath, BlobRange range,
+        Integer blockSize, ReliableDownloadOptions options, BlobAccessConditions accessConditions,
+        boolean rangeGetContentMD5) {
+        return withContext(context -> downloadToFileWithResponse(filePath, range, blockSize, options, accessConditions,
+                rangeGetContentMD5, context));
     }
 
-    Mono<Void> downloadToFile(String filePath, BlobRange range, Integer blockSize, ReliableDownloadOptions options,
-        BlobAccessConditions accessConditions, boolean rangeGetContentMD5, Context context) {
+    Mono<Response<BlobProperties>> downloadToFileWithResponse(String filePath, BlobRange range, Integer blockSize,
+                              ReliableDownloadOptions options, BlobAccessConditions accessConditions,
+                              boolean rangeGetContentMD5, Context context) {
         if (blockSize != null) {
             Utility.assertInBounds("blockSize", blockSize, 0, BLOB_MAX_DOWNLOAD_BLOCK_SIZE);
         }
 
         return Mono.using(() -> downloadToFileResourceSupplier(filePath),
-            channel -> Mono.justOrEmpty(range)
-                .switchIfEmpty(getFullBlobRange(accessConditions))
-                .flatMapMany(rg -> Flux.fromIterable(sliceBlobRange(rg, blockSize)))
-                .flatMap(chunk -> this.download(chunk, accessConditions, rangeGetContentMD5, context)
-                    .subscribeOn(Schedulers.elastic())
-                    .flatMap(dar -> FluxUtil.writeFile(dar.body(options), channel,
-                        chunk.getOffset() - (range == null ? 0 : range.getOffset()))))
-                .then(), this::downloadToFileCleanup);
+            channel -> blobDownloadInChunk(channel, range, blockSize, options, accessConditions, rangeGetContentMD5,
+                context),
+            this::downloadToFileCleanup);
+
+    }
+
+    private Mono<Response<BlobProperties>> blobDownloadInChunk(final AsynchronousFileChannel channel, BlobRange range,
+                              Integer blockSize, ReliableDownloadOptions options, BlobAccessConditions accessConditions,
+                              boolean rangeGetContentMD5, Context context) {
+        Mono<Response<BlobProperties>> asyncGetPropertiesResponse = getPropertiesWithResponse(accessConditions);
+        Mono<BlobRange> fullBlobRange = (range == null) ? getFullBlobRange(asyncGetPropertiesResponse)
+            : Mono.just(range);
+        return fullBlobRange.flatMapMany(rg -> Flux.fromIterable(sliceBlobRange(rg, blockSize)))
```

Comment:
Could we look at what `downloadWithResponse` is doing to see if we could put this all into a single stream?

Reply:
downloadWithProperties return DownloadAsyncResponse. Do we want to align with that?

Reply:
Have you considered the .concat operator?

Reply:
More on the stream structure not the response type.